h/api/models.py

```diff
@@ -166,6 +170,112 @@ def title(self):
             return ""
 
     @property
+    def filename(self):
+        if self.uri and self.uri.startswith("file://"):
+            return self.uri.split("/")[-1] or ""
+        else:
+            return ""
+
+    @property
+    def document_link(self):
+        """Return a link to this annotation's document.
+
+        Returns HTML strings like:
+
+          <a href="{href}" title="{title}">{link_text}</a> ({domain})
+
+        where:
+
+        - {href} is the uri of the annotated document,
+          if it has an http(s):// uri
+        - {title} is the title of the document.
+          If the document has no title then its uri will be used instead.
+          If it's a local file:// uri then only the filename part is used,
+          not the full path.
+        - {link_text} is the same as {title}, but truncated with &hellip; if
+          it's too long
+        - {domain} is the domain name of the document's uri without
+          the scheme (http(s)://) and www parts, e.g. "example.com".
+          If it's a local file:// uri then the filename is used as the domain.
+          If the domain is too long it is truncated with &hellip;.
+
+        The ({domain}) part will be missing if it wouldn't be any different
+        from the {link_text} part.
+
+        The href="{href}" will be missing if there's no http(s) uri to link to
+        for this annotation's document.
+
+        User-supplied values are escaped so the string is safe for raw
+        rendering (the returned string is actually a jinja2.Markup object and
+        won't be escaped by Jinja2 when rendering).
+
+        """
+        uri = jinja2.escape(self.uri) or ""
+
+        if uri.startswith("http://") or uri.startswith("https://"):
+            href = uri
+        else:
+            href = ""
+
+        if self.title:
+            title = jinja2.escape(self.title)
+            link_text = title
+            if uri.startswith("file://"):
+                domain = jinja2.escape(self.filename)
+            else:
+                domain = urlparse.urlparse(uri).hostname
+        else:
+            if uri.startswith("file://"):
+                title = urllib2.unquote(jinja2.escape(self.filename))
```

Comment:
Is that injection really possible? I'm escaping the title, and I have a test for it